pkg/pool-weighted/contracts/smart/InvestmentPool.sol

```diff
@@ -193,29 +287,156 @@ contract InvestmentPool is BaseWeightedPool, ReentrancyGuard {
         uint256 decimalsDifference = tokenState.decodeUint5(_DECIMAL_DIFF_OFFSET);
 
         return FixedPoint.ONE * 10**decimalsDifference;
     }
 
+    // Swap overrides - revert unless swaps are enabled
+
+    function _onSwapGivenIn(
+        SwapRequest memory swapRequest,
+        uint256 currentBalanceTokenIn,
+        uint256 currentBalanceTokenOut
+    ) internal view override returns (uint256) {
+        _require(getSwapEnabled(), Errors.SWAPS_DISABLED);
+
+        return super._onSwapGivenIn(swapRequest, currentBalanceTokenIn, currentBalanceTokenOut);
+    }
+
+    function _onSwapGivenOut(
+        SwapRequest memory swapRequest,
+        uint256 currentBalanceTokenIn,
+        uint256 currentBalanceTokenOut
+    ) internal view override returns (uint256) {
+        _require(getSwapEnabled(), Errors.SWAPS_DISABLED);
+
+        return super._onSwapGivenOut(swapRequest, currentBalanceTokenIn, currentBalanceTokenOut);
+    }
+
+    /**
+     * @dev Extend ownerOnly functions to include the LBP control functions
+     */
+    function _isOwnerOnlyAction(bytes32 actionId) internal view override returns (bool) {
+        return
+            (actionId == getActionId(InvestmentPool.setSwapEnabled.selector)) ||
+            (actionId == getActionId(InvestmentPool.updateWeightsGradually.selector)) ||
+            super._isOwnerOnlyAction(actionId);
+    }
+
+    function _onJoinPool(
+        bytes32 poolId,
+        address sender,
+        address recipient,
+        uint256[] memory balances,
+        uint256 lastChangeBlock,
+        uint256 protocolSwapFeePercentage,
+        uint256[] memory scalingFactors,
+        bytes memory userData
+    )
+        internal
+        override
+        returns (
+            uint256 bptAmountOut,
+            uint256[] memory amountsIn,
+            uint256[] memory dueProtocolFeeAmounts
+        )
+    {
+        // TokenInForExactBPTOut is not proportional - so fail immediately if swaps are disabled
+        _require(
+            getSwapEnabled() || userData.joinKind() != JoinKind.TOKEN_IN_FOR_EXACT_BPT_OUT,
+            Errors.DISPROPORTIONATE_JOIN_OR_EXIT
+        );
+
+        (bptAmountOut, amountsIn, dueProtocolFeeAmounts) = super._onJoinPool(
+            poolId,
+            sender,
+            recipient,
+            balances,
+            lastChangeBlock,
+            protocolSwapFeePercentage,
+            scalingFactors,
+            userData
+        );
+
+        // If the pool is paused, only allow proportional join
+        if (!getSwapEnabled()) {
+            // We know the bptOut from the join - so proportional exit with that bptOut should reproduce amountsIn,
+            // if the join was proportional
+            uint256[] memory amountsOut = WeightedMath._calcTokensOutGivenExactBptIn(
+                balances,
+                bptAmountOut,
+                totalSupply().add(bptAmountOut)
+            );
```

Comment:
I don't love the fact that this block implicitly depends on `balances` having been properly mutated during `super._onJoinPool()`. As far as I can tell, the underlying call to `BaseWeightedPool._invariantAfterJoin()` should update `balances` accordingly, but are we willing to make that assumption at this level? Even in `BasePool.sol`, you'll find this comment after the `_onJoinPool()` call:

```
// Note we no longer use `balances` after calling `_onJoinPool`, which may mutate it.
```

The assumption is that `_onJoinPool` _may_ mutate `balances` while performing computations (for gas savings), not that `balances` _must_ correctly reflect the post-join state at the end of the call. I'd be really curious to get @nventuro's thoughts on this. Even if this currently works, are we willing to bet that it always will? And that if we make changes such that it no longer works, we will remember this conversation and fix this (and/or the tests will catch it)?

Reply:
Yes, but I'm not sure what could be done about it other than repeat all the calculations, or extract it from the new lastInvariant (which makes the same assumptions). Could at least add documentation